### PR TITLE
Use 0x API for trades

### PR DESCRIPTION
**Pull request by ncitron**
# **Pull Request Template**

## **Type of Change**

*Select a category that relates to the content of your pull request (choose one).*

- [ ] Bug fix
- [x] New feature
- [ ] Documentation
- [x] Code Refactoring
- [ ] Update of code styling

&nbsp;

## **Does this request include a breaking change?**

*Is there any potential for this change to crash Index UI or make it unusable?*

- [ ] Yes
- [x] No

&nbsp;

## **Description**

Link to specification: https://docs.google.com/document/d/1V9Fmdn0JQwEyvqWB5toi-nhH8PgE9fwun18fezgaOeU/edit?usp=sharing 

&nbsp;

**Summary of Changes:**
- use 0x's API to route trades
- remove premium in order summary (it was misleading, as it was the slippage, not the premium)

## **Checklist**

*Please review and confirm each of the following before creating your pull request.*

- [x] I have reviewed the [Contribution Guidelines](https://github.com/SetProtocol/index-ui/blob/master/CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have checked my code and corrected any errors.
- [x] I have commented my code or used an effective naming scheme.
- [x] I have updated my repository to match the master at `SetProtocol/index-ui`.
- [x] I have included test data or screenshots that prove my fix is effective or that my feature works.

&nbsp;

*If all of the above are **true**, then you may mark your pull request as **ready for review**.*

### CODE REVIEW

File: src/components/BuySell/components/BuySellButton.tsx
Comment (0xModene) [inline src/components/BuySell/components/BuySellButton.tsx:9]: So everything goes through 0x now? Such tidy, much wow

Comment (ncitron) [reply src/components/BuySell/components/BuySellButton.tsx:9]: Yep! One step closer to migrating to V3...

**Review by controtie — CHANGES_REQUESTED** [review]
Can you leave some notes about the 0x API in this PR? Like what the payload pieces are, what it means for us to submit it directly with web3.transaction etc... It's hard to dig in and tell if the implementation is correct. We should have specced this out a bit more before starting work.

I think there's also some minor UI bugs like all currency options being highlighted:
image

And the please confirm your transaction message still being shown after the transaction's been submitted:
image

File: src/components/BuySell/components/BuySellButton.tsx
Comment (ncitron) [reply src/components/BuySell/components/BuySellButton.tsx:58]: Yeah the undefined case is to prevent it from displaying an approve tokens message when the user is not logged in

File: src/components/BuySell/components/MaxButton.tsx
Comment (ncitron) [reply src/components/BuySell/components/MaxButton.tsx:56]: The root of this problem is that the useBalances hook returns balances in displayable values, rather than their precise (1e18) values. We end up losing some precision when balances get rounded during this conversion. We should definitely, think about refactoring the useBalances hook to fix this in the future, but it is quite deeply integrated into other parts of the codebase, so it probably deserves a separate PR

File: src/components/BuySell/components/BuySellButton.tsx
Comment (controtie) [inline src/components/BuySell/components/BuySellButton.tsx:63]: very nice

File: src/components/BuySell/components/MaxButton.tsx
Comment (controtie) [reply src/components/BuySell/components/MaxButton.tsx:55]: Makes sense, can you write something to that effect in the comment so we have that context in the future? maybe mention we lose precision in `BalancesProvider` and this rounds down to account for that.

File: src/contexts/BuySell/BuySellProvider.tsx
Comment (controtie) [inline src/contexts/BuySell/BuySellProvider.tsx:10]: just `utils/zeroExUtils` should work here, similar to the bignumber import above

File: src/utils/coingeckoApi.ts
Comment (controtie) [inline src/utils/coingeckoApi.ts:56]: remove console.logs. Also do we need try/catch blocks in here? if the coingecko request fails how does that affect the rest of the flow?

File: src/utils/coingeckoApi.ts
Comment (controtie) [inline src/utils/coingeckoApi.ts:33]: Rename to `fetchCoingeckoTokenPrice`

File: src/utils/zeroExUtils.ts
Comment (controtie) [inline src/utils/zeroExUtils.ts:72]: can we add a comment in here detailing what this does? is it mostly formatting for people to read? does this affect the payload that we submit to 0x?

File: src/utils/zeroExUtils.ts
Comment (controtie) [inline src/utils/zeroExUtils.ts:98]: Add a TODO, we should decouple this in a future PR

**Review by controtie — CHANGES_REQUESTED** [review]
Nice work on the previous changes, I've left some more comments. Also check out your token pricing stuff, something seems to be off when selling:

image

File: src/contexts/BuySell/BuySellProvider.tsx
Comment (controtie) [inline src/contexts/BuySell/BuySellProvider.tsx:16]: I think this is a mistaken import?

File: src/contexts/BuySell/BuySellProvider.tsx
Comment (controtie) [inline src/contexts/BuySell/BuySellProvider.tsx:123]: What's this do?

File: src/utils/zeroExUtils.ts
Comment (controtie) [inline src/utils/zeroExUtils.ts:68]: display purposes*

File: .env.prod
Comment (controtie) [inline .env.prod:14]: Is this canonical? Or is this specific to 0x's API?

Comment (controtie) [reply .env.prod:14]: It seems like this address is only used in relation to 0x API. If it's not canonical we should label it as such. Maybe include it as a constant in `ethContractAddresses` instead of setting it as ENV

**Review by controtie — CHANGES_REQUESTED** [review]
Just a couple more minor fixes. Will do a thorough test once these are in

File: src/contexts/BuySell/BuySellProvider.tsx
Comment (ncitron) [reply src/contexts/BuySell/BuySellProvider.tsx:123]: This is part of the logic for getting to the transaction failed screen. It justs sets the promise result to "0" if a tx fails, which is then checked for and thrown when it does fail (being caught at the bottom of the try/catch that this is in

Comment (controtie) [reply src/contexts/BuySell/BuySellProvider.tsx]: In this case can we just `reject()` like on 121? would also remove the need for line 128

File: src/utils/coingeckoApi.ts
Comment (controtie) [inline src/utils/coingeckoApi.ts:45]: This pattern looks a bit weird, but I tested it and it seems OK.

File: src/contexts/Balances/Provider.tsx
Comment (controtie) [inline src/contexts/Balances/Provider.tsx:111]: Let's hold off on this fix actually.

**Review by controtie — CHANGES_REQUESTED** [review]
Last couple things.

**Review by controtie — APPROVED** [review]
LGTM!

File: src/utils/zeroExUtils.ts
Comment (justinkchen) [inline src/utils/zeroExUtils.ts:54]: @ncitron can you explain to me why isExactInput is what differentiates if we use sellAmount or buyAmount?

**Pull request merged**
